common/network-types/src/frame.rs

```diff
@@ -136,62 +160,53 @@ impl Ord for Segment<'_> {
     }
 }
 
-impl From<Segment<'_>> for Box<[u8]> {
-    fn from(value: Segment<'_>) -> Self {
-        let mut ret = Vec::with_capacity(3 * mem::size_of::<u16>() + value.data.len());
+impl From<Segment> for Vec<u8> {
+    fn from(value: Segment) -> Self {
+        let mut ret = Vec::with_capacity(Segment::HEADER_SIZE + value.data.len());
         ret.extend_from_slice(value.frame_id.to_be_bytes().as_ref());
         ret.extend_from_slice(value.seq_idx.to_be_bytes().as_ref());
         ret.extend_from_slice(value.seq_len.to_be_bytes().as_ref());
-        ret.extend_from_slice(value.data);
-        ret.into_boxed_slice()
+        ret.extend_from_slice(value.data.as_ref());
+        ret
     }
 }
 
-impl<'a> TryFrom<&'a [u8]> for Segment<'a> {
+impl TryFrom<&[u8]> for Segment {
     type Error = NetworkTypeError;
 
-    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
-        let (header, data) = value.split_at(3 * mem::size_of::<u16>());
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let (header, data) = value.split_at(Self::HEADER_SIZE);
         let segment = Segment {
-            frame_id: u16::from_be_bytes(header[0..2].try_into().map_err(|_| InvalidSegment)?),
-            seq_idx: u16::from_be_bytes(header[2..4].try_into().map_err(|_| InvalidSegment)?),
-            seq_len: u16::from_be_bytes(header[4..6].try_into().map_err(|_| InvalidSegment)?),
-            data,
+            frame_id: FrameId::from_be_bytes(header[0..4].try_into().map_err(|_| NetworkTypeError::InvalidSegment)?),
+            seq_idx: u16::from_be_bytes(header[4..6].try_into().map_err(|_| NetworkTypeError::InvalidSegment)?),
+            seq_len: u16::from_be_bytes(header[6..8].try_into().map_err(|_| NetworkTypeError::InvalidSegment)?),
+            data: data.into(),
         };
-        (segment.seq_idx < segment.seq_len)
+        (segment.frame_id > 0 && segment.seq_idx < segment.seq_len)
             .then_some(segment)
-            .ok_or(InvalidSegment)
-    }
-}
-
-impl<'a> Segment<'a> {
-    /// Identifies the entire segment sequence using the frame id and the length
-    /// of the sequence.
-    #[inline]
-    pub fn seq_id(&self) -> u32 {
-        (self.frame_id as u32) << 16 | self.seq_len as u32
+            .ok_or(NetworkTypeError::InvalidSegment)
     }
 }
 
 /// Rebuilds the [Frame] from [Segments](Segment).
 #[derive(Debug)]
-struct FrameBuilder<'a> {
-    frame_id: u16,
-    segments: Vec<OnceLock<&'a [u8]>>,
+struct FrameBuilder {
+    frame_id: FrameId,
+    segments: Vec<OnceLock<Box<[u8]>>>,
```

Comment:
Just curious, what was the reason for removing the referential behavior and including the owning logic with `Box`?

Reply:
Impossible lifetime when implemented with AsyncRead and AsyncWrite 😒 

This is somewhat slower than the referential approach. Maybe I will revisit it later again.